[BUG] Python: lifespan pools race condition during graceful shutdown

**Describe the issue**
During graceful shutdown (`SIGTERM`), `_aio_start` and `exit_gracefully` race to clean up the `lifespan`, causing connection pools to be closed while tasks are still running. This results in connection pool errors as existing, running tasks try to acquire the connection.

I believe the race was introduced in PR #2908 which moved `_cleanup_lifespan` into `_aio_start` after the health check completes. The problem is that `_check_listener_health` exits its loop as soon as `self.killing` is set to `True`, which happens at the start of `exit_gracefully` — before `wait_for_tasks` has finished.

**Environment**
- SDK: 1.27.0
- Engine: Cloud

**Expected behavior**
Connection pools should remain open until all in-flight tasks have completed. `_cleanup_lifespan` should only run after `wait_for_tasks` finishes.

**Code to Reproduce, Logs, or Screenshots**
Any worker using a lifespan with connection pools will likely hit this on `SIGTERM` if tasks are in-flight, but I made a script that samples this. You need to BYOKey:

```
import asyncio
import os
from collections.abc import AsyncGenerator
from datetime import timedelta

from hatchet_sdk import Context, EmptyModel, Hatchet

hatchet = Hatchet()


class FakePool:
    """Simulates an async resource (e.g. a DB connection pool) that can be closed."""

    def __init__(self):
        self._closed = False

    def acquire(self):
        if self._closed:
            raise RuntimeError("the pool is already closed")
        print(f"  [pool] connection acquired")

    def close(self):
        self._closed = True
        print(f"  [pool] closed")

    @property
    def closed(self):
        return self._closed


async def lifespan() -> AsyncGenerator[FakePool, None]:
    pool = FakePool()
    print(f"[lifespan] Pool opened (pid={os.getpid()})")
    try:
        yield pool
    finally:
        pool.close()
        print(f"[lifespan] Pool torn down (pid={os.getpid()})")


@hatchet.task(
    name="repro.slow_task",
    execution_timeout=timedelta(minutes=5),
)
async def slow_task(_: EmptyModel, ctx: Context) -> dict:
    pool: FakePool = ctx.lifespan
    for i in range(30):
        print(f"[task] iteration {i}, acquiring connection...")
        pool.acquire()  # Will raise "pool is already closed" if lifespan cleaned up early
        print(f"[task] iteration {i} done, sleeping 1s...")
        await asyncio.sleep(1)
    return {"done": True}


if __name__ == "__main__":
    worker = hatchet.worker(
        "repro-shutdown-race",
        workflows=[slow_task],
        lifespan=lifespan,
        slots=5,
    )
    print(f"Worker PID: {os.getpid()}")
    worker.start()
```

Once starting this worker you then:
1. Create a `repro.slow_task` task
2. After a moment to let the iterations begin, you can send a `kill -SIGTERM <pid>`

Eventually it yields something like this depending on when the `SIGTERM` actually happens

```
...
[task] iteration 3 done, sleeping 1s...
[INFO]  🪓 -- 2026-03-04 19:10:05,847 - received signal SIGTERM...
[INFO]  🪓 -- 2026-03-04 19:10:06,360 - waiting for 1 tasks to finish...
[task] iteration 4, acquiring connection...
  [pool] connection acquired
[task] iteration 4 done, sleeping 1s...
  [pool] closed
[lifespan] Pool torn down (pid=84055)
[INFO]  🪓 -- 2026-03-04 19:10:07,361 - waiting for 1 tasks to finish...
[task] iteration 5, acquiring connection...
[ERROR] 🪓 -- 2026-03-04 19:10:07,364 - failed step run: repro.slow_task:repro.slow_task/ec9f9b85-6d04-4d05-ab3c-60a1d295a611
RuntimeError: the pool is already closed
...
```

--- a/sdks/python/CHANGELOG.md
+++ b/sdks/python/CHANGELOG.md
@@ -5,6 +5,12 @@ All notable changes to Hatchet's Python SDK will be documented in this changelog
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [1.28.1] - 2026-03-05
+
+### Changed
+
+- Fixes a bug where lifespans are shut down eagerly before the worker is drained, causing unexpected behavior in still-running tasks.
+
 ## [1.28.0] - 2026-03-02
 
 ### Added

--- a/sdks/python/examples/lifespans/drain.py
+++ b/sdks/python/examples/lifespans/drain.py
@@ -1,0 +1,68 @@
+import asyncio
+from collections.abc import AsyncGenerator
+from dataclasses import dataclass
+
+from hatchet_sdk import Context, Hatchet
+
+hatchet = Hatchet()
+
+
+class FakePool:
+    def __init__(self) -> None:
+        self._closed = False
+
+    def acquire(self) -> None:
+        if self._closed:
+            raise RuntimeError("the pool is already closed")
+
+    def close(self) -> None:
+        self._closed = True
+
+    @property
+    def closed(self) -> bool:
+        return self._closed
+
+
+async def drain_lifespan() -> AsyncGenerator[FakePool, None]:
+    pool = FakePool()
+    try:
+        yield pool
+    finally:
+        pool.close()
+
+
+@dataclass
+class DrainResult:
+    status: str
+    iterations: int
+
+
+@dataclass
+class DrainInput:
+    n: int
+
+
+@hatchet.task(name="LifespanDrainTask", input_validator=DrainInput)
+async def lifespan_drain_task(input: DrainInput, ctx: Context) -> DrainResult:
+    pool: FakePool = ctx.lifespan
+    iters = 0
+    for _ in range(input.n):
+        pool.acquire()
+        iters += 1
+        await asyncio.sleep(1)
+
+    return DrainResult(status="ok", iterations=iters)
+
+
+def main() -> None:
+    worker = hatchet.worker(
+        "drain-test-worker",
+        slots=5,
+        workflows=[lifespan_drain_task],
+        lifespan=drain_lifespan,
+    )
+    worker.start()
+
+
+if __name__ == "__main__":
+    main()

--- a/sdks/python/examples/lifespans/test_lifespans.py
+++ b/sdks/python/examples/lifespans/test_lifespans.py
@@ -1,6 +1,15 @@
+import asyncio
+import signal
+from subprocess import Popen
+from typing import Any
+
+import psutil
 import pytest
 
+from examples.lifespans.drain import lifespan_drain_task, DrainInput
 from examples.lifespans.simple import Lifespan, lifespan_task
+from hatchet_sdk import Hatchet
+from hatchet_sdk import RunStatus
 
 
 @pytest.mark.asyncio(loop_scope="session")
@@ -10,3 +19,47 @@ async def test_lifespans() -> None:
     assert isinstance(result, Lifespan)
     assert result.pi == 3.14
     assert result.foo == "bar"
+
+
+@pytest.mark.parametrize(
+    "on_demand_worker",
+    [
+        (
+            ["poetry", "run", "python", "examples/lifespans/drain.py"],
+            8009,
+        )
+    ],
+    indirect=True,
+)
+@pytest.mark.asyncio(loop_scope="session")
+async def test_lifespan_drain_on_sigterm(
+    hatchet: Hatchet,
+    on_demand_worker: Popen[Any],
+) -> None:
+    n = 6
+    ref = await lifespan_drain_task.aio_run_no_wait(DrainInput(n=n))
+
+    for _ in range(30):
+        run = await hatchet.runs.aio_get_details(ref.workflow_run_id)
+        if run.status == RunStatus.RUNNING:
+            break
+
+        await asyncio.sleep(1)
+    else:
+        pytest.fail("Task never entered RUNNING state")
+
+    await asyncio.sleep(1)
+
+    parent = psutil.Process(on_demand_worker.pid)
+    children = parent.children(recursive=True)
+
+    for proc in [parent] + children:
+        try:
+            proc.send_signal(signal.SIGTERM)
+        except psutil.NoSuchProcess:
+            pass
+
+    result = await ref.aio_result()
+
+    assert result.status == "ok"
+    assert result.iterations == n

--- a/sdks/python/hatchet_sdk/deprecated/worker.py
+++ b/sdks/python/hatchet_sdk/deprecated/worker.py
@@ -142,8 +142,13 @@ async def legacy_aio_start(worker: Worker) -> None:
             )
         )
 
-        await worker.action_runner.wait_for_tasks()
         await worker.action_listener_health_check
+
+        if worker.action_runner:
+            await worker.action_runner.wait_for_tasks()
+
+        if durable_action_runner:
+            await durable_action_runner.wait_for_tasks()
 
         try:
             await worker._cleanup_lifespan()

--- a/sdks/python/hatchet_sdk/deprecated/worker.py
+++ b/sdks/python/hatchet_sdk/deprecated/worker.py
@@ -142,6 +142,7 @@ async def legacy_aio_start(worker: Worker) -> None:
             )
         )
 
+        await worker.action_runner.wait_for_tasks()
         await worker.action_listener_health_check
 
         try:

--- a/sdks/python/hatchet_sdk/worker/worker.py
+++ b/sdks/python/hatchet_sdk/worker/worker.py
@@ -299,8 +299,7 @@ class Worker:
 
             await self.action_listener_health_check
 
-            if self.action_runner:
-                await self.action_runner.wait_for_tasks()
+            await self.action_runner.wait_for_tasks()
 
             try:
                 await self._cleanup_lifespan()

--- a/sdks/python/hatchet_sdk/worker/worker.py
+++ b/sdks/python/hatchet_sdk/worker/worker.py
@@ -299,6 +299,9 @@ class Worker:
 
             await self.action_listener_health_check
 
+            if self.action_runner:
+                await self.action_runner.wait_for_tasks()
+
             try:
                 await self._cleanup_lifespan()
             except LifespanSetupError:
@@ -348,7 +351,9 @@ class Worker:
     async def _cleanup_lifespan(self) -> None:
         try:
             if self.lifespan_stack is not None:
-                await self.lifespan_stack.aclose()
+                stack = self.lifespan_stack
+                self.lifespan_stack = None
+                await stack.aclose()
         except Exception as e:
             logger.exception("error during lifespan cleanup")
             raise LifespanSetupError("An error occurred during lifespan cleanup") from e

--- a/sdks/python/pyproject.toml
+++ b/sdks/python/pyproject.toml
@@ -1,6 +1,6 @@
 [tool.poetry]
 name = "hatchet-sdk"
-version = "1.28.0"
+version = "1.28.1"
 description = "This is the official Python SDK for Hatchet, a distributed, fault-tolerant task queue. The SDK allows you to easily integrate Hatchet's task scheduling and workflow orchestration capabilities into your Python applications."
 authors = [
   "Alexander Belanger <alexander@hatchet.run>",

